classes/uki-esp-image.bbclass

```diff
@@ -0,0 +1,59 @@
+# Create an ESP image that has an UKI, systemd-boot and the matching configs
+
+# Optional subfolder, dependant on where the ESP partition gets mounted
+# intended to only have a leading slash, no trailing slash e.g. '/EFI', or just empty, ''
+ESPFOLDER ?= "/EFI"
+
+# Optional initrd, filname inside ${DEPLOY_DIR_IMAGE}
+ESPINITRD ?= ""
+
+# Title as shown on the console
+UKI_TITLE ?= "${DISTRO_NAME} ${DISTRO_VERSION} - ${MACHINE}"
+
+# Kernel CMDLINE
+UKI_CMDLINE = " "
+UKI_CMDLINE[export] = "1"
+
+do_ukiesp() {
+	mkdir -p ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries
+
+	# Copy over files from deploy into the rootfs
+	install -m 0755 ${DEPLOY_DIR_IMAGE}/${UKI_FILENAME} ${IMAGE_ROOTFS}${ESPFOLDER}
+	install -m 0644 ${DEPLOY_DIR_IMAGE}/${KERNEL_DEVICETREE} ${IMAGE_ROOTFS}${ESPFOLDER}
+
+	# Add global settings to systemd-boot config
+	echo "timeout 5" > ${IMAGE_ROOTFS}${ESPFOLDER}/loader/loader.conf
+
+	# copy over kernel and optional initrd and dtb
+	echo "title ${UKI_TITLE}" > ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
+	echo "linux /${UKI_FILENAME}" >> ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
+
+	# Assemble CMDLINE
+	# FIXME: don't use QCOM_* variables
+	if [ -n "${QCOM_BOOTIMG_ROOTFS}" ]; then
+	    UKI_CMDLINE="$UKI_CMDLINE root=${QCOM_BOOTIMG_ROOTFS} rw rootwait"
+	fi
+	if [ -n "${SERIAL_CONSOLES}" ]; then
+	    tmp="${SERIAL_CONSOLES}"
+	    for entry in $tmp ; do
+		baudrate=`echo $entry | sed 's/\;.*//'`
+		tty=`echo $entry | sed -e 's/^[0-9]*\;//' -e 's/\;.*//'`
+		console="$tty","$baudrate"n8
+		UKI_CMDLINE="$UKI_CMDLINE console=$console"
+	    done
+	fi
+	if [ -n "${KERNEL_CMDLINE_EXTRA}" ]; then
+	    UKI_CMDLINE="$UKI_CMDLINE ${KERNEL_CMDLINE_EXTRA}"
+	fi
+	echo "options $UKI_CMDLINE" >> ${IMAGE_ROOTFS}${ESPFOLDER}/loader/entries/boot.conf
```

Comment:
Does this override or extend the command line in the UKI?

Reply:
From a build perspective, it will extend what is passed into UKI_CMDLINE. From a boot perspective, the order seems to be:

1. The provided cmdline in boot.conf _replaces_
2. The provided cmdline inside the UKI _replaces_
3.  A compiled-in cmdline in the *Image

The systemd-boot documentation is ambiguous, the most definitive discussions seems to be here: 1.  A compiled-in cmdline in the *Image

Reply:
This behavior depends if secure boot is enabled as well.